### PR TITLE
All halide_mutex and halide_cond instances should be explicitly zero-inited

### PR DESCRIPTION
**Pull request by steven-johnson**

### CODE REVIEW

**Comment by abadams** [comment]
Any particular reason this came up, or is this a style thing? I usually rely on globals being zero-inited to avoid accidentally invoking a static constructor.

**Comment by abadams** [comment]
(I've had problems in the past with initializers on statics requiring cxa_lock_guard junk because C++11 guarantees they're atomic)

**Comment by steven-johnson** [comment]
Style thing. Does the runtime actually guarantee that globals will be zero-inited?

File: src/runtime/thread_pool_common.h [hunk 1]
```diff
@@ -57,7 +57,18 @@ struct work_queue_t {
     }
 
 };
-WEAK work_queue_t work_queue;
+WEAK work_queue_t work_queue = {
```
Comment (abadams) [inline src/runtime/thread_pool_common.h]: Would in-class initializers work here to avoid specifying all the fields twice?

**Comment by zvookin** [comment]
Globals are guaranteed to be initialized to 0, or to another constant if specified by some means such as an aggregate initializer or a default value in the class decl.

The only difference resulting from putting in an initializer to zero should be that the line becomes a definition unambiguously. Otherwise it could be a forward decl. Though it may be that this has been cleaned up in C++ since the last time I read the spec in detail. (And maybe this only applies to things declared extern, though I don't think so.)

I don't think the existence of an initializer should change whether there is a lock guard or not *unless* the initializer requires a runtime computed value.

I general, we should steer toward data types ensuring their own initialization and not rely on them being filled in at the point of use.

**Comment by steven-johnson** [comment]
> Globals are guaranteed to be initialized to 0

In that case, this whole CL is pointless. Withdrawing.

**Pull request closed without merging**